maitake/src/wait/queue.rs

```diff
@@ -0,0 +1,790 @@
+                State::Waiting => break,
+                _ => {}
+            }
+
+            let next = state.with_state(State::Woken);
+            // advance the state to `Woken`, and return (if we did so
+            // successfully)
+            match self.compare_exchange(state, next) {
+                Ok(_) => return,
+                Err(actual) => state = actual,
+            }
+        }
+
+        // okay, there are tasks waiting on the queue; we must acquire the lock
+        // on the linked list and wake the next task from the queue.
+        let mut queue = self.queue.lock();
+        test_trace!("wake: -> locked");
+
+        // the queue's state may have changed while we were waiting to acquire
+        // the lock, so we need to acquire a new snapshot.
+        state = self.load();
+
+        if let Some(waker) = self.wake_locked(&mut *queue, state) {
+            drop(queue);
+            waker.wake();
+        }
+    }
+
+    /// Wake *all* tasks currently in the queue.
+    pub fn wake_all(&self) {
+        let mut queue = self.queue.lock();
+        let state = self.load();
+
+        match state.get(QueueState::STATE) {
+            // if the queue is closed, bail.
+            State::Closed => return,
+
+            // if there are no waiters in the queue, increment the number of
+            // `wake_all` calls and return.
+            State::Woken | State::Empty => {
+                self.state.fetch_add(QueueState::ONE_WAKE_ALL, SeqCst);
+                return;
+            }
+            State::Waiting => {}
+        }
+
+        // okay, we actually have to wake some stuff.
+
+        // TODO(eliza): wake outside the lock using an array, a la
+        // https://github.com/tokio-rs/tokio/blob/4941fbf7c43566a8f491c64af5a4cd627c99e5a6/tokio/src/sync/batch_semaphore.rs#L277-L303
+        while let Some(node) = queue.pop_back() {
+            let waker = Waiter::wake(node, &mut queue, Wakeup::All);
+            waker.wake()
+        }
+
+        // now that the queue has been drained, transition to the empty state,
+        // and increment the wake_all count.
+        let next_state = QueueState::new()
+            .with_state(State::Empty)
+            .with(QueueState::WAKE_ALLS, state.get(QueueState::WAKE_ALLS) + 1);
+        self.compare_exchange(state, next_state)
+            .expect("state should not have transitioned while locked");
+    }
+
+    /// Close the queue, indicating that it may no longer be used.
+    ///
+    /// Once a queue is closed, all [`wait`] calls (current or future) will
+    /// return an error.
+    ///
+    /// This method is generally used when implementing higher-level
+    /// synchronization primitives or resources: when an event makes a resource
+    /// permanently unavailable, the queue can be closed.
+    pub fn close(&self) {
+        let state = self.state.fetch_or(State::Closed as u8 as usize, SeqCst);
```

Comment:
Might wanna put a big ol' warning flag on the State variable that this is dependent on "Closed" being "all bits set" of the relevant state fields